tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -48,35 +63,69 @@
+   * @param firstShortInProgress earliest in-progress short transaction
+   * @param type transaction type
+   * @param checkpointPointers the list of writer pointers added from checkpoints on the transaction
+   */
+  public Transaction(long readPointer, long writePointer, long currentWritePointer, long[] invalids, long[] inProgress,
+                     long firstShortInProgress, TransactionType type, long[] checkpointPointers) {
     this.readPointer = readPointer;
     this.writePointer = writePointer;
+    this.currentWritePointer = currentWritePointer;
     this.invalids = invalids;
     this.inProgress = inProgress;
     this.firstShortInProgress = firstShortInProgress;
     this.type = type;
+    this.checkpointWritePointers = checkpointPointers;
   }
 
   /**
-   * Creates a new short transaction.
-   * @param readPointer read pointer for transaction
-   * @param writePointer write pointer for transaction
-   * @param invalids list of invalid transactions to exclude while reading
-   * @param inProgress list of in-progress transactions to exclude while reading
-   * @param firstShortInProgress earliest in-progress short transaction
+   * Creates a new transaction for a checkpoint operation, copying all members from the original transaction,
+   * with the updated checkpoint write pointers.
+   *
+   * @param toCopy the original transaction containing the state to copy
+   * @param currentWritePointer the new write pointer to use for the transaction
+   * @param checkpointPointers the list of write pointers added from checkpoints on the transaction
    */
-  public Transaction(long readPointer, long writePointer, long[] invalids, long[] inProgress,
-                     long firstShortInProgress) {
-    this(readPointer, writePointer, invalids, inProgress, firstShortInProgress, TransactionType.SHORT);
+  public Transaction(Transaction toCopy, long currentWritePointer, long[] checkpointPointers) {
+    this(toCopy.getReadPointer(), toCopy.getWritePointer(), currentWritePointer, toCopy.getInvalids(),
+        toCopy.getInProgress(), toCopy.getFirstShortInProgress(), toCopy.getType(), checkpointPointers);
   }
 
   public long getReadPointer() {
     return readPointer;
   }
 
+  /**
+   * Returns the initial write pointer assigned to the transaction.  This will remain the same for the life of the
+   * transaction, and uniquely identifies it with the transaction service.  This value should be provided
+   * to identify the transaction when calling any transaction lifecycle methods on the transaction service.
+   */
   public long getWritePointer() {
     return writePointer;
   }
 
+  /**
+   * Returns the write pointer to be used in persisting any changes.  After a checkpoint is performed, this will differ
+   * from {@link #getWritePointer()}.  This method should always be used when setting the timestamp for writes
+   * in order to ensure that the correct value is used.
+   */
+  public long getCurrentWritePointer() {
```

Comment:
I don't see how it is backward incompatible to use transactionID and writePointer. For any transaction client that existed before checkpoint, the transaction id and the write pointer will always be the same (because they don't enable checkpointing). 

I think currentWritePointer can be confusing, too, if we ever extend to allow a single transaction to have multiple active write pointers at the same time. (like we would for MapReduce if we assign a different write pointer to each mapper and reducer). 

Semantically (and from the description of this PR), it is more natural to call the "writePointer (same through the life of the transaction, used to identify the tx in TransactionManager)" the transactionId. Whereas currentWritePointer should be named writePointer. That is backward-compatible because existing clients call getWritePointer() to obtain the version with which to write their changes.